### ISSUE TITLE
MPI support

### ISSUE THREAD

**Issue by paugier** [conda-forge/fftw-feedstock#51]
It is more a question than a proper issue for this feedstock.

I'd need fftw3_mpi in conda-forge. Similar to these Debian package:

- https://packages.debian.org/stretch/libfftw3-mpi
- https://packages.debian.org/stretch/libfftw3-mpi-dev

These packages depend on libfftw3 and openmpi. They do not contain the libfftw3 libraries.

It would be nice to have the same structure on conda, i.e. that the conda packages fftw and fftw_mpi would be compatible (fftw would be a dependence of fftw_mpi).

The code for fftwmpi-feedstock would be very similar than the one for this feedstock and it would also produce the same libraries as this recipe, so I'm wondering if it's a good thing to have another feedstock or if it would be possible with one feedstock to create two packages.

**Comment by beckermr** [comment]
A few options.

1. We could have a nompi option similar to hdf5. 
2. We should check howFFTW works. Maybe all of the libraries can coexist?

If we had a separate feedstock, we would only want to ship the MPI libs or make it incompatible with this one if they have to ship the same libs.

**Comment by grlee77** [comment]
I have not used the MPI functionality, but I am pretty sure these get built as separate libraries in addition to the current ones.
I think aside from adding the MPI dependency in `meta.yaml` we just need to add `--enable-mpi` in the build scripts to get these variants.

See:
http://www.fftw.org/fftw3_doc/FFTW-MPI-Installation.html

**Comment by grlee77** [comment]
Perhaps @rainwoodman might have a better idea as I think he has recent experience with MPI and FFTW

**Comment by grlee77** [comment]
Based on a past PR that was never completed for pyFFTW, it seems that the library names for MPI are indeed different and have the same precision suffixes as the serial case:
e.g. `libfftw3-mpi`, `libfftw3f-mpi`, `libfftw3l-mpi` and there is a separate `fftw3-mpi.h` header corresponding to these.

**Comment by beckermr** [comment]
Are there changes in the symbols in the other libraries?

**Comment by grlee77** [comment]
@beckermr
For hdf5 does the version without `nompi` provide both serial and MPI variants along side each other or is it MPI only?
For FFTW it looks like both variants can coexist together. I think the main benefit of a `nompi` here would just be allowing the dependency on MPI to be removed and give a smaller download for users who aren't interested in that functionality.



> Are there changes in the symbols in the other libraries?

As far as I can tell, it looks like all functions have `_mpi` in the function name, so the symbols do not conflict with those in the serial library.  e.g. `fftw_mpi_execute_dft` instead of `fftw_execute_dft`
see:
https://github.com/FFTW/fftw3/blob/master/mpi/fftw3-mpi.h#L201-L203

**Comment by beckermr** [comment]
IDK about hdf5. 

I think actually more than one output in this recipe is the right way to do this. 

We build everything here in this feedstock. Then we supply two outputs.

1. an fftw-mpi package that has the dependence on the MPI provider and the MPI headers and SO's. 2. an fftw output that holds everything else

**Comment by grlee77** [comment]
> We build everything here in this feedstock. Then we supply two outputs.

That sounds reasonable. I didn't realize that was a possibility. I looked it up and found the following issue that links to an example recipe using the approach: in https://github.com/conda-forge/conda-forge.github.io/issues/746

**Comment by beckermr** [comment]
Yep. This is documented in the anaconda docs as well. I can do this, but not until next week or so.

**Comment by beckermr** [comment]
OK. Since the mpi libs can simply be installed along side everything else, I have instead decided to configure things like hdf5. There will be a preferred nompi package, but mpi-enabled variants can be installed. Packages build against the nompi variant can link to the mpi variants at runtime as well.

**Comment by rainwoodman** [comment]
The configuration is a bit strange -- the mpi / no mpi does not feel like a variant. The mpi version may depend on the nompi version though.

I'd suggest splitting into two packages, fftw3 and fftw3-mpi, then treating mpich/openmpi as variants of the mpi package, either of which depends on fftw3.

This way the MPI package shall only ship with files for the MPI support (libfftw3-mpi.* and the mpi header). I know that conda supports generating subpackages, though I am not very familiar with the syntax.

```
fftw3
fftw3-mpi[openmpi]
fftw3-mpi[mpich]
```

**Comment by beckermr** [comment]
> The configuration is a bit strange -- the mpi / no mpi does not feel like a variant. The mpi version may depend on the nompi version though.

The mpi versions ship all of the normal files plus the mpi headers and libs, so there is no dependence between them.

> I'd suggest splitting into two packages, fftw3 and fftw3-mpi, then treating mpich/openmpi as variants of the mpi package, either of which depends on fftw3.
> This way the MPI package shall only ship with files for the MPI support (libfftw3-mpi.* and the mpi header). 

Given the docs for MPI in conda-forge and what has been done for hdf5, we probably have to keep it this way. See the docs http://conda-forge.org/docs/maintainer/knowledge_base.html#message-passing-interface-mpi.

> I know that conda supports generating subpackages, though I am not very familiar with the syntax.

I know this syntax and thought this originally at first too. However, given the packaging of hdf5, it makes more sense to do things this way. 

The final bit here is that FFTW is not setup to have the MPI installation split out into separate packages. The installer is made to do it all at once (even though it does ship separate headers/libs). However this is an implementation detail of the upstream package. In the long term, mucking around here to split the monolithic build is not good practice. If we want a split build, we should make an issue upstream.

**Comment by rainwoodman** [comment]
The conda-forge policy may be problematic. 

Consider the scenario if a user needs a package A that depends on nompi and a package B that depends on mpich at the same time -- the policy seems to say the A and B cannot be co-installed in the same environment, even though they actually should be able to co-exist.

There are plenty of prior experience to follow. On Fedora, both packages are split into mpi and nonmpi sub-packages. The FFTW mpi subpackage depends on the nompi subpackage, while the HDF mpi subpackage doesn't. What about debian, gentoo, and ArchLinux?

```
============================================================== Name Matched: fftw* ===============================================================
fftw.x86_64 : A Fast Fourier Transform library
fftw-devel.x86_64 : Headers, libraries and docs for the FFTW library
fftw-static.x86_64 : Static versions of the FFTW libraries
fftw-mpich-devel.x86_64 : Headers, libraries and docs for the FFTW MPICH library
fftw-mpich-static.x86_64 : Static versions of the FFTW MPICH libraries
fftw-openmpi-devel.x86_64 : Headers, libraries and docs for the FFTW OpenMPI library
fftw-openmpi-static.x86_64 : Static versions of the FFTW OpenMPI libraries

========================================================= Summary & Name Matched: hdf5* ==========================================================
hdf5-mpich.x86_64 : HDF5 mpich libraries
hdf5-devel.x86_64 : HDF5 development files
hdf5-static.x86_64 : HDF5 static libraries
hdf5-devel.x86_64 : HDF5 development files
hdf5-static.x86_64 : HDF5 static libraries
hdf5-openmpi.x86_64 : HDF5 openmpi libraries
hdf5-mpich-devel.x86_64 : HDF5 mpich development files
hdf5-mpich-static.x86_64 : HDF5 mpich static libraries
hdf5-openmpi-devel.x86_64 : HDF5 openmpi development files
hdf5-openmpi-static.x86_64 : HDF5 openmpi static libraries
```

**Comment by beckermr** [comment]
I don't make the rules. Raise this issue on the main conda-forge repos if you feel strongly.

The way this is addressed now is in the docs starting around 

> This doesn’t extend to nompi

and the build number bump and the lack of a run export for the nompi version.

For any package built against nompi fftw, the current build allows an mpi version to be used at run time. Thus for this build it is fine.

**Comment by rainwoodman** [comment]
Thanks for the explanation. Now I see how it works, and the uneasiness is no longer as strong.  I am OK to  assume there are no packages where the nompi ABI and the nompi portion of the mpi ABI are incompatible. The magic number 100 still stinks, but not bad enough for changing the rule.
